r/src/r_to_arrow.cpp

```diff
@@ -0,0 +1,1045 @@
+
+class RStructConverter : public StructConverter<RConverter, RConverterTrait> {
+ public:
+  Status Extend(SEXP x, int64_t size) override {
+    // check that x is compatible
+    R_xlen_t n_columns = XLENGTH(x);
+
+    if (!Rf_inherits(x, "data.frame") && !Rf_inherits(x, "POSIXlt")) {
+      return Status::Invalid("Can only convert data frames to Struct type");
+    }
+
+    auto fields = this->struct_type_->fields();
+    if (n_columns != static_cast<R_xlen_t>(fields.size())) {
+      return Status::RError("Number of fields in struct (", fields.size(),
+                            ") incompatible with number of columns in the data frame (",
+                            n_columns, ")");
+    }
+
+    cpp11::strings x_names = Rf_getAttrib(x, R_NamesSymbol);
+
+    RETURN_NOT_OK(cpp11::unwind_protect([&] {
+      for (int i = 0; i < n_columns; i++) {
+        const char* name_i = arrow::r::unsafe::utf8_string(x_names[i]);
+        auto field_name = fields[i]->name();
+        if (field_name != name_i) {
+          return Status::RError(
+              "Field name in position ", i, " (", field_name,
+              ") does not match the name of the column of the data frame (", name_i, ")");
+        }
+      }
+
+      return Status::OK();
+    }));
+
+    for (R_xlen_t i = 0; i < n_columns; i++) {
+      std::string name(x_names[i]);
+      if (name != fields[i]->name()) {
+        return Status::RError(
+            "Field name in position ", i, " (", fields[i]->name(),
+            ") does not match the name of the column of the data frame (", name, ")");
+      }
+    }
+
+    for (R_xlen_t i = 0; i < n_columns; i++) {
+      SEXP x_i = VECTOR_ELT(x, i);
+      if (vctrs::short_vec_size(x_i) < size) {
+        return Status::RError("Degenerated data frame");
+      }
+    }
+
+    RETURN_NOT_OK(this->Reserve(size));
+
+    for (R_xlen_t i = 0; i < size; i++) {
+      RETURN_NOT_OK(struct_builder_->Append());
+    }
+
+    for (R_xlen_t i = 0; i < n_columns; i++) {
+      auto status = children_[i]->Extend(VECTOR_ELT(x, i), size);
+      if (!status.ok()) {
+        return Status::Invalid("Problem with column ", (i + 1), " (", fields[i]->name(),
+                               "): ", status.ToString());
+      }
+    }
+
+    return Status::OK();
+  }
+
+ protected:
+  Status Init(MemoryPool* pool) override {
+    return StructConverter<RConverter, RConverterTrait>::Init(pool);
+  }
+};
+
+template <>
+struct RConverterTrait<DictionaryType> {
+  template <typename T>
+  using dictionary_type = RDictionaryConverter<T>;
+};
+
+// ---- short circuit the Converter api entirely when we can do zero-copy
+
+// in some situations we can just use the memory of the R object in an RBuffer
+// instead of going through ArrayBuilder, etc ...
+bool can_reuse_memory(SEXP x, const std::shared_ptr<arrow::DataType>& type) {
+  // TODO: this probably should be disabled when x is an ALTREP object
+  //       because MakeSimpleArray below will force materialization
```

Comment:
Isn't this unavoidable though? We have to get the data into Arrow format somehow.

I guess the one case you might not want to materialize is if you had an ALTREP object that were backed by an Arrow array (which of course we don't do yet)--then you could just grab the Arrow array.

Reply:
Maybe that's a reach, but perhaps there are cases where this would mean only arrow materialize without R materialize. For example if the arrow data type is not representable in R, e.g. `uint32`